lb.tf

```
@@ -15,76 +15,157 @@ resource "cloudscale_server_group" "lb" {
-  name             = "${random_id.lb[count.index].hex}.${local.node_name_suffix}"
-  zone_slug        = "${var.region}1"
-  flavor_slug      = "plus-8"
-  image_slug       = "ubuntu-20.04"
-  server_group_ids = var.lb_count != 0 ? [cloudscale_server_group.lb[0].id] : []
-  volume_size_gb   = 50
-  ssh_keys         = var.ssh_keys
+  count                          = var.lb_count
+  name                           = local.instance_fqdns[count.index]
+  zone_slug                      = "${var.region}1"
+  flavor_slug                    = "plus-8"
+  image_slug                     = "ubuntu-20.04"
+  server_group_ids               = var.lb_count != 0 ? [cloudscale_server_group.lb[0].id] : []
+  volume_size_gb                 = 50
+  ssh_keys                       = var.ssh_keys
+  skip_waiting_for_ssh_host_keys = true
   interfaces {
     type = "public"
   }
   interfaces {
     type         = "private"
     network_uuid = cloudscale_network.privnet.id
-    no_address   = true
   }
   lifecycle {
     create_before_destroy = true
   }
-  user_data = <<-EOF
-    #cloud-config
-    package_update: true
-    packages:
-    - haproxy
-    - keepalived
-    bootcmd:
-    - "iptables -t nat -A POSTROUTING -o ens3 -j MASQUERADE"
-    - "sysctl -w net.ipv4.ip_forward=1"
-    - "sysctl -w net.ipv4.ip_nonlocal_bind=1"
-    - "ip link set ens7 up"
-    - "ip address add ${cidrhost(var.privnet_cidr, 2 + count.index)}/24 dev ens7"
-    write_files:
-    - path: "/etc/keepalived/keepalived.conf"
-      encoding: b64
-      content: ${base64encode(templatefile("${path.module}/templates/keepalived.conf", {
-  "api_eip" = random_id.lb[count.index].keepers.api_eip
-  "api_int" = cidrhost(var.privnet_cidr, 100)
-  "gateway" = cloudscale_subnet.privnet_subnet.gateway_address
-  "prio"    = "${(var.lb_count - count.index) * 10}"
-  }))}
-    - path: "/etc/haproxy/haproxy.cfg"
-      encoding: b64
-      content: ${base64encode(templatefile("${path.module}/templates/haproxy.cfg", {
-  "api_eip" = cidrhost(cloudscale_floating_ip.api_vip[0].network, 0)
-  "api_int" = cidrhost(var.privnet_cidr, 100)
-  "api_servers" = [
-    cidrhost(var.privnet_cidr, 10),
-    "etcd-0.${local.node_name_suffix}",
-    "etcd-1.${local.node_name_suffix}",
-    "etcd-2.${local.node_name_suffix}",
-  ]
-  "infra_servers" = length(random_id.lb[count.index].keepers.infra_servers) > 0 ? split(",", random_id.lb[count.index].keepers.infra_servers) : []
-}))}
-    EOF
-}
 
-resource "null_resource" "api_vip_assignement" {
-  count = var.lb_count != 0 ? 1 : 0
-  triggers = {
-    api_eip = cloudscale_floating_ip.api_vip[0].network
-    server  = cloudscale_server.lb[0].id
-  }
+  user_data = format("#cloud-config\n%s", yamlencode(merge(
+    local.common_user_data,
+    {
+      "fqdn"             = local.instance_fqdns[count.index],
+      "hostname"         = random_id.lb[count.index].hex,
+      "manage_etc_hosts" = true,
+    },
+    {
+      "write_files" = local.common_write_files,
+    }
```

These can be merged into a single object:

```suggestion
    {
      "fqdn"             = local.instance_fqdns[count.index],
      "hostname"         = random_id.lb[count.index].hex,
      "manage_etc_hosts" = true,
      "write_files"      = local.common_write_files,
    }
```

Also, since we don't need the ec2_metadata fixing here, the `write_files` could be moved to the local variable `common_user_data`.